code/controllers/Processes/explosives.dm

```diff
@@ -0,0 +1,200 @@
+								//It will return 0 if the sound volume falls to 0 due to falloff or pressure
+								//Also return zero if sound playing failed for some other reason
+
+						//Deaf people will feel vibrations though
+						if (volume > 0)//Only shake camera if someone was close enough to hear it
+							shake_camera(M, min(60,max(2,(power*18) / dist)), min(3.5,((power*3) / dist)),0.05)
+							//Maximum duration is 6 seconds, and max strength is 3.5
+							//Becuse values higher than those just get really silly
+
+		if(adminlog)
+			message_admins("Explosion with size ([devastation_range], [heavy_impact_range], [light_impact_range]) in area [epicenter.loc.name] ([epicenter.x],[epicenter.y],[epicenter.z]) (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[epicenter.x];Y=[epicenter.y];Z=[epicenter.z]'>JMP</a>)")
+			log_game("Explosion with size ([devastation_range], [heavy_impact_range], [light_impact_range]) in area [epicenter.loc.name] ")
+
+		var/approximate_intensity = (devastation_range * 3) + (heavy_impact_range * 2) + light_impact_range
+		powernet_rebuild_was_deferred_already = defer_powernet_rebuild
+		// Large enough explosion. For performance reasons, powernets will be rebuilt manually
+		if(!defer_powernet_rebuild && (approximate_intensity > 25))
+			defer_powernet_rebuild = 1
+
+		if(heavy_impact_range > 1)
+			var/datum/effect/system/explosion/E = new/datum/effect/system/explosion()
+			E.set_up(epicenter)
+			E.start()
+
+		var/x0 = epicenter.x
+		var/y0 = epicenter.y
+		var/z0 = epicenter.z
+
+		for(var/turf/T in trange(max_range, epicenter))
+			var/dist = sqrt((T.x - x0)**2 + (T.y - y0)**2)
+
+			if(dist < devastation_range)		dist = 1
+			else if(dist < heavy_impact_range)	dist = 2
+			else if(dist < light_impact_range)	dist = 3
+			else								continue
+
+			T.ex_act(dist)
+			SCHECK
+			if(T)
+				for(var/atom_movable in T.contents)	//bypass type checking since only atom/movable can be contained by turfs anyway
+					var/atom/movable/AM = atom_movable
+					if(AM && AM.simulated)	AM.ex_act(dist)
+					SCHECK
+
+		var/took = (world.timeofday-start)/10
+		//You need to press the DebugGame verb to see these now....they were getting annoying and we've collected a fair bit of data. Just -test- changes  to explosion code using this please so we can compare
+		if(Debug2)	world.log << "## DEBUG: Explosion([x0],[y0],[z0])(d[devastation_range],h[heavy_impact_range],l[light_impact_range]): Took [took] seconds."
+
+		//Machines which report explosions.
+		for(var/i,i<=doppler_arrays.len,i++)
+			var/obj/machinery/doppler_array/Array = doppler_arrays[i]
+			if(Array)
+				Array.sense_explosion(x0,y0,z0,devastation_range,heavy_impact_range,light_impact_range,took)
+
+		work_queue -= data
+
+	if(!powernet_rebuild_was_deferred_already && defer_powernet_rebuild)
+		makepowernets()
+		defer_powernet_rebuild = 0
+
+
+/datum/controller/process/explosives/proc/queue(var/data)
```

Comment:
Maybe just in case make this run istype() and !data checks? That way you ensure in the integrity of work_queue, without bogging down the main loop.